File: backend/src/main/java/org/jdk/project/controller/KnowledgeController.java
package org.jdk.project.controller;

import lombok.RequiredArgsConstructor;
import org.jdk.project.dto.ListResponse;
import org.jdk.project.dto.knowledge.KnowledgeUpsertRequest;
import org.jooq.generated.project.tables.pojos.KnowledgeBase;
import org.jdk.project.service.KnowledgeService;
import org.springframework.web.bind.annotation.DeleteMapping;
import org.springframework.web.bind.annotation.GetMapping;
import org.springframework.web.bind.annotation.PathVariable;
import org.springframework.web.bind.annotation.PostMapping;
import org.springframework.web.bind.annotation.PutMapping;
import org.springframework.web.bind.annotation.RequestBody;
import org.springframework.web.bind.annotation.RequestMapping;
import org.springframework.web.bind.annotation.RestController;

@RestController
@RequestMapping("/knowledge")
@RequiredArgsConstructor
public class KnowledgeController {

  private final KnowledgeService knowledgeService;

  @GetMapping
  public ListResponse<KnowledgeBase> listKnowledge() {
    return knowledgeService.listKnowledge();
  }

  @GetMapping("/{id}")
  public KnowledgeBase getKnowledge(@PathVariable Long id) {
    return knowledgeService.getKnowledge(id);
  }

  @PostMapping
  public Long createKnowledge(@RequestBody KnowledgeUpsertRequest request) {
    return knowledgeService.createKnowledge(request);
  }

  @PutMapping("/{id}")
  public void updateKnowledge(@PathVariable Long id, @RequestBody KnowledgeUpsertRequest request) {
    knowledgeService.updateKnowledge(id, request);
  }

  @DeleteMapping("/{id}")
  public void deleteKnowledge(@PathVariable Long id) {
    knowledgeService.deleteKnowledge(id);
  }
}

File: backend/src/main/kotlin/org/jdk/project/controller/KnowledgeController.kt
package org.jdk.project.controller

import org.jdk.project.dto.ListResponse
import org.jdk.project.dto.knowledge.KnowledgeUpsertRequest
import org.jdk.project.service.KnowledgeService
import org.jooq.generated.project.tables.pojos.KnowledgeBase
import org.springframework.web.bind.annotation.DeleteMapping
import org.springframework.web.bind.annotation.GetMapping
import org.springframework.web.bind.annotation.PathVariable
import org.springframework.web.bind.annotation.PostMapping
import org.springframework.web.bind.annotation.PutMapping
import org.springframework.web.bind.annotation.RequestBody
import org.springframework.web.bind.annotation.RequestMapping
import org.springframework.web.bind.annotation.RestController

@RestController
@RequestMapping("/knowledge")
class KnowledgeController(
    private val knowledgeService: KnowledgeService,
) {
    @GetMapping
    fun listKnowledge(): ListResponse<KnowledgeBase> = knowledgeService.listKnowledge()

    @GetMapping("/{id}")
    fun getKnowledge(@PathVariable id: Long): KnowledgeBase = knowledgeService.getKnowledge(id)

    @PostMapping
    fun createKnowledge(@RequestBody request: KnowledgeUpsertRequest): Long? = knowledgeService.createKnowledge(request)

    @PutMapping("/{id}")
    fun updateKnowledge(@PathVariable id: Long, @RequestBody request: KnowledgeUpsertRequest) {
        knowledgeService.updateKnowledge(id, request)
    }

    @DeleteMapping("/{id}")
    fun deleteKnowledge(@PathVariable id: Long) {
        knowledgeService.deleteKnowledge(id)
    }
}
